AI #1793: Create Commitment Program Eligibility Details column (#1882)
FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec: supporting_content/datasets/cost_and_usage/columns/commitmenteligibilitydetails.md

Reviewer: This examples belong directly within the column definition, not supporting content.  Look at the current definition for [AllocatedMethodDetails](https://github.com/FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec/blob/working_draft/specification/datasets/cost_and_usage/columns/allocatedmethoddetails.md) and [ContractApplied](https://github.com/FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec/blob/working_draft/specification/datasets/cost_and_usage/columns/contractapplied.md), and use that pattern.

Author: Addressed. Please review and let me know if we require changes

Reviewer: Not addressed, unfortunately.  This content is still in a separate markdown within `supporting_content`.  This content belongs in the column definition itself.  If you leave it where it is, it doesn't follow our pattern, but more importantly, it won't be in the spec.  `supporting_content` is our working group development scratchpad.

Here are the examples I previously linked back on the 15th, which have now been replaced by the refactor implemented via #2007, which was merged this past Thu -- so here are some fresh links:

https://github.com/FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec/blob/9fec2c1c17c403a24c1ec64ae43f6aecc451726e/specification/datasets/cost_and_usage/columns/contractapplied.md
https://github.com/FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec/blob/63db0db9c8e0e843bd6c122772d7ae98e6f33aba/specification/datasets/cost_and_usage/columns/allocatedmethoddetails.md

If/when this PR is approved and merged, then per agreement with the Maintainers, and as I [previously mentioned](https://github.com/FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec/pull/1882#issuecomment-4071075236): we will refactor this PR's content to follow the format as proposed via #2007, as we did not want to block your progress/approval with a fairly meaty administrative task.  But for the time being, you are obligated to get the content in the right spot per the old pattern.

Please reach out to @Matt-Cowsert if you need assistance, as he is obviously committed to helping you get this PR across the finish line.

Reviewer: Looks good now.  Resolving!

Reviewer: @SanjnaSrivatsaVMware 
@Matt-Cowsert has added a calculation for Coverage in the supported feature for this column PR #2025.

We should definitely add one or more concrete examples here, demonstrating how that calculation would work with hard numbers.

Could you partner with @Matt-Cowsert to either augment some of the examples here, or add a couple at the end?

Reviewer: @SanjnaSrivatsaVMware @Matt-Cowsert 
Something like this (note: AI-generated, do not just copy/paste).  Hopefully it gets the point across for what to _do_ with this column, not just show how it might be formatted.

### Example Coverage Calculations

The primary value of the `CommitmentEligibilityDetails` column is adjusting the denominator of a commitment coverage calculation to exclude ineligible spend. This provides a highly actionable coverage metric.

**Example 1: Compute and Storage (Provider A)**

Consider the following sample usage data for Provider A:

| ServiceProviderName | ServiceName | ChargeCategory | EffectiveCost | CommitmentDiscountId | CommitmentEligibilityDetails |
|:---|:---|:---|:---|:---|:---|
| Provider A | Virtual Machines | Usage | 10.00 | `arn:providera:savingsplans...` | `{"CommitmentPrograms": [{"ProgramType": "SavingsPlan"}]}` |
| Provider A | Virtual Machines | Usage | 15.00 | null | `{"CommitmentPrograms": [{"ProgramType": "SavingsPlan"}]}` |
| Provider A | ObjectStorage | Usage | 25.00 | null | null |

By isolating the denominator to only rows where `CommitmentEligibilityDetails` is not null, the $25.00 of ineligible ObjectStorage spend is correctly excluded:

* **Eligibility-Adjusted Coverage:** 10.00 Covered / 25.00 Total Eligible Spend = **40% Coverage**

**Example 2: Observability Platform (Provider B)**

Consider the following sample usage data for Provider B, where infrastructure hosts are eligible for monthly commitments, but custom metrics are not:

| ServiceProviderName | ServiceName | ChargeCategory | EffectiveCost | CommitmentDiscountId | CommitmentEligibilityDetails |
|:---|:---|:---|:---|:---|:---|
| Provider B | Infrastructure | Usage | 100.00 | `sub_12345` | `{"CommitmentPrograms": [{"ProgramType": "MonthlyCommitment"}]}` |
| Provider B | Infrastructure | Usage | 200.00 | null | `{"CommitmentPrograms": [{"ProgramType": "MonthlyCommitment"}]}` |
| Provider B | Custom Metrics | Usage | 50.00 | null | null |

Excluding the $50.00 of ineligible Custom Metrics spend yields the true coverage rate for eligible services:

* **Eligibility-Adjusted Coverage:** 100.00 Covered / 300.00 Total Eligible Spend = **33.3% Coverage**

Reviewer: Resolving in favor of a newly-posted thread, for visibility.

# Column: CommitmentEligibilityDetails

## Example

The following scenarios illustrate how CommitmentEligibilityDetails is populated across different providers, including major cloud platforms and SaaS data platforms.

| Provider | Service | ChargeClass | CommitmentEligibilityDetails |
|----------|---------|-------------|------------------------------|
| AWS | AmazonEC2 | Usage | {"EligibleCommitmentTypes": ["SavingsPlan", "ReservedInstance"]} |
| Azure | Virtual Machines | Usage | {"EligibleCommitmentTypes": ["SavingsPlan"]} |
| Google | Kubernetes Engine | Usage | {"EligibleCommitmentTypes": ["CommittedUseDiscount"]} |
| Oracle | Virtual Machine | Usage | {"EligibleCommitmentTypes": ["UniversalCredits"]} |
| Snowflake | Warehouse | Usage | {"EligibleCommitmentTypes": ["CapacityCommitment"]} |
| Databricks | Jobs | Usage | {"EligibleCommitmentTypes": ["CommittedUseDiscount"]} |
| AWS | AmazonS3 | Usage | null |

> ## Example usage scenarios

Reviewer 2: I'd add clear column headers for readability, similar to the example above in lines 7-8

Current values observed in billing data for various scenarios:

**Scenario 1: AWS (On-Demand EC2 Usage)**
An On-Demand EC2 instance that is not currently covered by any commitment but is eligible for both Savings Plan and Reserved Instance.
`AWS` | `AmazonEC2` | `Usage` | `null` | `{"EligibleCommitmentTypes": ["SavingsPlan", "ReservedInstance"]}`

**Scenario 2: Azure (Partially covered Virtual Machine Usage)**
A Virtual Machine usage row that is partially covered by Savings Plan.
`Azure` | `Virtual Machines` | `Usage` | `null` | `{"EligibleCommitmentTypes": ["SavingsPlan", "ReservedInstance"]}`

**Scenario 3: Google Cloud (Fully covered GKE Usage)**
A GKE cluster usage row that is covered by a Resource-based Committed Use Discount (CUD), but could also have been covered by a Compute Flexible CUD.
`Google` | `Kubernetes Engine` | `Usage` | `null` | `{"EligibleCommitmentTypes": ["ResourceBasedCommittedUseDiscount", "ComputeFlexibleCommittedUseDiscount"]}`

> **Scenario 4: Oracle Cloud Infrastructure (Standard Rates)**

Reviewer 2: Oracle "Universal Credits" scenario needs clarification. Universal Credits can apply to virtually any OCI service, so the eligibility logic might be too broad. Add guidance on how granular this should be. @ijurica may have a perspective we can lean on here.

Author: I agree. I am not sure how the other commitment related columns handle this. Do we consider Sustained Use Discount as a Commitment Discount Type also?

Reviewer 3: OCI Universal credits fall into the negotiated bucket

OCI Compute usage that is charged at standard rates but is eligible to burn down a "Universal Credits" commitment contract.
`Oracle` | `Virtual Machine` | `Usage` | `null` | `{"EligibleCommitmentTypes": ["UniversalCredits"]}`

> **Scenario 5: Datadog (Monthly and Annual Commitment)**

Reviewer 2: It may be worth adding a note here stating that custom commitment types like 'MonthlyCommitment' should be documented by the provider and consistently used in CommitmentDiscountType.

Author: I've tried to make that very clear in the spec. Let me know if there are specifc sentences that are still not clear

# Column: CommitmentEligibilityDetails

## Example

The following scenarios illustrate how CommitmentEligibilityDetails are populated across different providers, including major cloud platforms and SaaS data platforms.

| ServiceProviderName | ServiceName       | ChargeClass | CommitmentEligibilityDetails                                                                                                  |
|--------------|--------------|--------------|-----------------|

> | AWS                 | AmazonEC2         | Usage       | {"CommitmentDiscountTypes": [{"Type": "SavingsPlan"}, {"Type": "ReservedInstance"}]}                                          |

Reviewer 2: ```suggestion
| AWS                 | AmazonEC2         | Usage       | {"CommitmentPrograms": [{"ProgramType": "SavingsPlan"}, {"ProgramType": "ReservedInstance"}]}                                          |
```

> | Azure               | Virtual Machines  | Usage       | {"CommitmentDiscountTypes": [{"Type": "SavingsPlan"}, {"Type": "ReservedInstance"}]}                                          |

Reviewer 2: ```suggestion
| Azure               | Virtual Machines  | Usage       | {"CommitmentPrograms": [{"ProgramType": "SavingsPlan"}, {"ProgramType": "ReservedInstance"}]}                                          |
```

> | Google              | Kubernetes Engine | Usage       | {"CommitmentDiscountTypes": [{"Type": "ResourceBasedCommittedUseDiscount"}, {"Type": "ComputeFlexibleCommittedUseDiscount"}]} |

Reviewer 2: ```suggestion
| Google              | Kubernetes Engine | Usage       | {"CommitmentPrograms": [{"ProgramType": "ResourceBasedCommittedUseDiscount"}, {"ProgramType": "ComputeFlexibleCommittedUseDiscount"}]} |
```

> | Snowflake           | Warehouse         | Usage       | {"CommitmentDiscountTypes": [{"Type": "CapacityCommitment"}]}                                                                 |

Reviewer 2: ```suggestion
| Snowflake           | Warehouse         | Usage       | {"CommitmentPrograms": [{"ProgramType": "CapacityCommitment"}]}                                                                 |
```

> | Databricks          | Jobs              | Usage       | {"CommitmentDiscountTypes": [{"Type": "CommittedUseDiscount"}]}                                                               |

Reviewer 2: ```suggestion
| Databricks          | Jobs              | Usage       | {"CommitmentPrograms": [{"ProgramType": "CommittedUseDiscount"}]}                                                               |
```

> | Datadog             | Infrastructure    | Usage       | {"CommitmentDiscountTypes": [{"Type": "MonthlyCommitment"}, {"Type": "AnnualCommitment"}]}                                    |

Reviewer 2: ```suggestion
| Datadog             | Infrastructure    | Usage       | {"CommitmentPrograms": [{"ProgramType": "MonthlyCommitment"}, {"ProgramType": "AnnualCommitment"}]}                                    |
```

| AWS                 | AmazonS3          | Usage       | null                                                                                                                          |

## Example usage scenarios

**AWS (On-Demand EC2 Usage)**

> An On-Demand EC2 instance that is not currently covered by any commitment but is eligible for both Savings Plan and Reserved Instance.

Reviewer 2: ```suggestion
An On-Demand EC2 instance that is not currently covered by any commitment but is eligible for Savings Plan, Reserved Instance, and capacity reservations.
```

| ServiceProviderName | ServiceName | ChargeClass | CommitmentEligibilityDetails                                                         |
|---------------|---------------|---------------|---------------|

> | AWS                 | AmazonEC2   | Usage       | {"CommitmentDiscountTypes": [{"Type": "SavingsPlan"}, {"Type": "ReservedInstance"}]} |

Reviewer 2: ```suggestion
| AWS                 | AmazonEC2   | Usage       | {"CommitmentPrograms": [{"ProgramType": "SavingsPlan"}, {"ProgramType": "ReservedInstance"}, {"ProgramType": "CapacityReservation"}]} |
```

Adds a capacity reservation ProgramType to show the mixed-type case.

**Google Cloud (Fully covered GKE Usage)**

A GKE cluster usage row that is covered by a Resource-based Committed Use Discount (CUD), but could also have been covered by a Compute Flexible CUD.

| ServiceProviderName | ServiceName       | ChargeClass | CommitmentEligibilityDetails                                                                                                  |
|---------------|---------------|---------------|---------------|

> | Google              | Kubernetes Engine | Usage       | {"CommitmentDiscountTypes": [{"Type": "ResourceBasedCommittedUseDiscount"}, {"Type": "ComputeFlexibleCommittedUseDiscount"}]} |

Reviewer 2: ```suggestion
| Google              | Kubernetes Engine | Usage       | {"CommitmentPrograms": [{"ProgramType": "ResourceBasedCommittedUseDiscount"}, {"ProgramType": "ComputeFlexibleCommittedUseDiscount"}]} |
```

**Databricks**

Databricks Units (DBUs) consumed by a Jobs cluster. This usage is eligible for coverage under a Databricks Committed Use Discount plan.

| ServiceProviderName | ServiceName | ChargeClass | CommitmentEligibilityDetails                                    |
|---------------|---------------|---------------|---------------|

> | Databricks          | Jobs        | Usage       | {"CommitmentDiscountTypes": [{"Type": "CommittedUseDiscount"}]} |

Reviewer 2: ```suggestion
| Databricks          | Jobs        | Usage       | {"CommitmentPrograms": [{"ProgramType": "CommittedUseDiscount"}]} |
```

**Ineligible Usage**

Standard S3 Storage usage or a support fee, which is not eligible for any commitment program.

| ServiceProviderName | ServiceName | ChargeClass | CommitmentEligibilityDetails |
|---------------|---------------|---------------|---------------|
| AWS                 | AmazonS3    | Usage       | null                         |

**Negotiated commitment (provider opts in)**

OCI Compute usage eligible for Universal Credits. Because Universal Credits are negotiated, providers MAY include them at their discretion. This example shows a provider that chooses to include them.

| ServiceProviderName | ServiceName     | ChargeClass | CommitmentEligibilityDetails                                |
|---------------|---------------|---------------|---------------|

> | Oracle              | Virtual Machine | Usage       | {"CommitmentDiscountTypes": [{"Type": "UniversalCredits"}]} |

Reviewer 2: ```suggestion
| Oracle              | Virtual Machine | Usage       | {"CommitmentPrograms": [{"ProgramType": "UniversalCredits"}]} |
```

# Column: CommitmentEligibilityDetails

## Example

> The following scenarios illustrate how CommitmentEligibilityDetails is populated across different providers, including major cloud platforms and SaaS data platforms.

Reviewer 4: ```suggestion
The following scenarios illustrate how CommitmentEligibilityDetails are populated across different providers, including major cloud platforms and SaaS data platforms.
```

# Column: CommitmentEligibilityDetails

## Example

> The following scenarios illustrate how CommitmentEligibilityDetails are populated across different providers, including major cloud platforms and SaaS data platforms.

Reviewer: I'm concerned about the framing here, using service providers that, at present, do not generate any FOCUS data.  For example:

> Since Datadog does not populate CommitmentDiscountType, the Type values correspond to publicly available program names from Datadog documentation.

While I fully understand the desire to ground the reader's understanding in real-world providers, this is dangerous territory for you as the author (and thus for the spec in general) unless you have independently verified that each of these examples is true according to a real-world FOCUS implementation.

I strongly recommend you convert these to generic service provider names.

Reviewer: Looks like you have simplified to Providers A through D.  Resolving.

| AWS        | AmazonEC2         | Compute            | Usage       | {"CommitmentDiscountTypes": [{"Type": "SavingsPlan"}, {"Type": "ReservedInstance"}]}                                          |
| Azure      | Virtual Machines  | VirtualMachine     | Usage       | {"CommitmentDiscountTypes": [{"Type": "SavingsPlan"}, {"Type": "ReservedInstance"}]}                                          |
| Google     | Kubernetes Engine | Node               | Usage       | {"CommitmentDiscountTypes": [{"Type": "ResourceBasedCommittedUseDiscount"}, {"Type": "ComputeFlexibleCommittedUseDiscount"}]} |
| Snowflake  | Warehouse         | n/a                | Usage       | {"CommitmentDiscountTypes": [{"Type": "CapacityCommitment"}]}                                                                 |
| Databricks | Jobs              | n/a                | Usage       | {"CommitmentDiscountTypes": [{"Type": "CommittedUseDiscount"}]}                                                               |
| Datadog    | Infrastructure    | Host               | Usage       | {"CommitmentDiscountTypes": [{"Type": "MonthlyCommitment"}, {"Type": "AnnualCommitment"}]}                                    |

> | AWS        | AmazonS3          | StandardStorage    | Usage       | null                                                                                                                          |

Reviewer: This overview table includes Service and ServiceSubcategory.

1) It's `ServiceName`, not `Service`.  (This is a change you need to apply globally across all examples.)
2) It's `ServiceProviderName`, not `Provider`.  (This is a change you need to apply globally across all examples.)
3) You are not referencing `ServiceSubcategory` across the examples below.  I'd recommend just striking it from this table.

Author: Done. These were very weird inconsistencies. Sorry about that. I check for names and tried to keep them consistent in my latest commit

Author: I'll let you review and respond if this now looks better. I'm still working through other comments. This is specific to what is mentioned in this comment.

Reviewer: Confirmed fixed.  Resolving!

## Example usage scenarios

**Scenario 1: AWS (On-Demand EC2 Usage)**

An On-Demand EC2 instance that is not currently covered by any commitment but is eligible for both Savings Plan and Reserved Instance.

| Provider | Service   | ChargeClass | CommitmentDiscountStatus | CommitmentEligibilityDetails                                                         |
|---------------|---------------|---------------|---------------|---------------|
| AWS      | AmazonEC2 | Usage       | null                     | {"CommitmentDiscountTypes": [{"Type": "SavingsPlan"}, {"Type": "ReservedInstance"}]} |

**Scenario 2: Azure (Partially covered Virtual Machine Usage)**

A Virtual Machine usage row that is partially covered by Savings Plan. The eligibility column still reflects all programs this usage qualifies for, regardless of current coverage.

| Provider | Service          | ChargeClass | CommitmentDiscountStatus | CommitmentEligibilityDetails                                                         |
|---------------|---------------|---------------|---------------|---------------|

> | Azure    | Virtual Machines | Usage       | null                     | {"CommitmentDiscountTypes": [{"Type": "SavingsPlan"}, {"Type": "ReservedInstance"}]} |

Reviewer: The description states this is a "Partially covered" Virtual Machine usage row. However, the table lists the CommitmentDiscountStatus as null. If it is partially covered, the status should reflect a covered state (e.g., Used).

Reviewer 3: In this use case, we would need two rows, right? One with the CommitmentDiscountStatus set to Used, and one without, i.e., set to null, right?

Attaching the relevant DiscountHandling requirements for context:

> * All discounts applied to a row MUST apply to the entire charge.
>   ...
>   * If a discount only applies to a portion of a charge, then the discounted **portion of the charge MUST be split into a separate row**.
>   * Each discount MUST be identifiable using existing FOCUS columns.
>     * Rows with a *commitment discount* applied to them MUST include a CommitmentDiscountId.

Reviewer: All use of Commitment Discount Status has been struck from the examples.  Resolving.

**Scenario 3: Google Cloud (Fully covered GKE Usage)**

A GKE cluster usage row that is covered by a Resource-based Committed Use Discount (CUD), but could also have been covered by a Compute Flexible CUD.

| Provider | Service           | ChargeClass | CommitmentDiscountStatus | CommitmentEligibilityDetails                                                                                                  |
|---------------|---------------|---------------|---------------|---------------|

> | Google   | Kubernetes Engine | Usage       | null                     | {"CommitmentDiscountTypes": [{"Type": "ResourceBasedCommittedUseDiscount"}, {"Type": "ComputeFlexibleCommittedUseDiscount"}]} |

Reviewer: The description states this is a "Fully covered" GKE usage row. Like Scenario 2, the table lists the CommitmentDiscountStatus as null.  I believe this should be `Used`.

Reviewer: Looks like you've struck CommitmentDiscountStatus from the examples.  Resolving.

|---------------|---------------|---------------|---------------|---------------|
| Databricks | Jobs    | Usage       | null                     | {"CommitmentDiscountTypes": [{"Type": "CommittedUseDiscount"}]} |

**Scenario 6: Ineligible Usage**

Standard S3 Storage usage or a support fee, which is not eligible for any commitment program.

| Provider | Service  | ChargeClass | CommitmentDiscountStatus | CommitmentEligibilityDetails |
|---------------|---------------|---------------|---------------|---------------|
| AWS      | AmazonS3 | Usage       | null                     | null                         |

**Scenario 7: Negotiated commitment (provider opts in)**

OCI Compute usage eligible for Universal Credits. Because Universal Credits are negotiated, providers MAY include them at their discretion. This example shows a provider that chooses to include them.

| Provider | Service         | ChargeClass | CommitmentDiscountStatus | CommitmentEligibilityDetails                                |
|---------------|---------------|---------------|---------------|---------------|

> | Oracle   | Virtual Machine | Usage       | null                     | {"CommitmentDiscountTypes": [{"Type": "UniversalCredits"}]} |

Reviewer 3: ```suggestion
| Oracle   | COMPUTE | Usage       | null                     | {"CommitmentDiscountTypes": [{"Type": "UniversalCredits"}]} |
```

**Scenario 8: AWS (Capacity Reservation-eligible EC2 Usage)**

> An EC2 instance type and tenancy that are eligible for both Savings Plans/Reserved Instances and for capacity reservations (for example, regional or zonal reservations). The eligibility column reflects all commitment constructs the usage qualifies for.

Reviewer: ```suggestion
An EC2 instance type and tenancy that are eligible for both Savings Plans/Reserved Instances and for capacity reservations (e.g., regional reservations, zonal reservations). The eligibility column reflects all commitment constructs the usage qualifies for.
```

# Column: CommitmentEligibilityDetails

## Example

The following scenarios illustrate how CommitmentEligibilityDetails are populated across different providers, including major cloud platforms and SaaS data platforms.

> | Provider   | Service           | ChargeClass | CommitmentEligibilityDetails                                                                                                  |

Reviewer 5: You need to add ServiceSubcategory column as for example AmazonEC2 itself isn't eligible but Compute instance(part of EC2) is eligible all other parts of EC2 are not.

Author: addressed

| Databricks | Jobs    | Usage       | null                     | {"CommitmentDiscountTypes": [{"Type": "CommittedUseDiscount"}]} |

**Scenario 6: Ineligible Usage**

Standard S3 Storage usage or a support fee, which is not eligible for any commitment program.

| Provider | Service  | ChargeClass | CommitmentDiscountStatus | CommitmentEligibilityDetails |
|---------------|---------------|---------------|---------------|---------------|
| AWS      | AmazonS3 | Usage       | null                     | null                         |

**Scenario 7: Negotiated commitment (provider opts in)**

OCI Compute usage eligible for Universal Credits. Because Universal Credits are negotiated, providers MAY include them at their discretion. This example shows a provider that chooses to include them.

| Provider | Service         | ChargeClass | CommitmentDiscountStatus | CommitmentEligibilityDetails                                |
|---------------|---------------|---------------|---------------|---------------|
| Oracle   | Virtual Machine | Usage       | null                     | {"CommitmentDiscountTypes": [{"Type": "UniversalCredits"}]} |

> **Scenario 8: AWS (Capacity Reservation–eligible EC2 Usage)**

Reviewer 2: ```suggestion
**Scenario 8: AWS (Capacity Reservation-eligible EC2 Usage)**
```

En-dash (U+2013) replaced with hyphen. Will fail the MD990 linter otherwise.

Author: addressed
